servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/AfterFinallyHttpOperator.java

```diff
@@ -0,0 +1,440 @@
+import io.servicetalk.concurrent.api.Publisher;
+import io.servicetalk.concurrent.api.Single;
+import io.servicetalk.concurrent.api.SingleOperator;
+import io.servicetalk.concurrent.api.TerminalSignalConsumer;
+import io.servicetalk.concurrent.internal.CancelImmediatelySubscriber;
+import io.servicetalk.http.api.StreamingHttpRequest;
+import io.servicetalk.http.api.StreamingHttpResponse;
+
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+import javax.annotation.Nullable;
+
+import static io.servicetalk.concurrent.api.SourceAdapters.toSource;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.atomic.AtomicIntegerFieldUpdater.newUpdater;
+
+/**
+ * Helper operator for signaling the end of an HTTP Request/Response cycle.
+ *
+ * <p>{@link StreamingHttpRequest} and {@link StreamingHttpResponse} are nested sources ({@link Single} of meta-data
+ * containing a payload {@link Publisher}), which makes it non-trivial to get a single signal at the end of this
+ * Request/Response cycle. One needs to consider and coordinate between the multitude of outcomes: cancel/success/error
+ * across both sources.</p>
+ * <p>This operator ensures that the provided callback is triggered just once whenever the sources reach a terminal
+ * state across both sources. An important question is when the ownership of the callback is transferred from the
+ * {@link Single} source and the payload {@link Publisher}. In this case ownership is transferred the first time the
+ * payload is subscribed to. This means that if a cancellation of the response {@link Single} occurs after the response
+ * has been emitted but before the message body has been subscribed to, the callback will observe a cancel. However, if
+ * the message payload has been subscribed to, the cancellation of the {@link Single} will have no effect and the result
+ * is dictated by the terminal event of the payload body. If the body is subscribed to multiple times, only the first
+ * subscribe will receive ownership of the terminal events.</p>
+ *
+ * Example usage tracking the begin and end of a request:
+ *
+ * <pre>{@code
+ *     // coarse grained, any terminal signal calls the provided `Runnable`
+ *     return requester.request(strategy, request)
+ *                     .beforeOnSubscribe(__ -> tracker.requestStarted())
+ *                     .liftSync(new AfterFinallyHttpOperator(tracker::requestFinished));
+ *
+ *     // fine grained, `tracker` implements `TerminalSignalConsumer`, terminal signal indicated by the callback method
+ *     return requester.request(strategy, request)
+ *                     .beforeOnSubscribe(__ -> tracker.requestStarted())
+ *                     .liftSync(new AfterFinallyHttpOperator(tracker));
+ * }</pre>
+ */
+public final class AfterFinallyHttpOperator implements SingleOperator<StreamingHttpResponse, StreamingHttpResponse> {
```

Comment:
To nobodies surprise, there is a very large amount of duplication in here with the `BeforeFinallyHttpOperator`. We could consider consolidating it into a `WhenFinallyHttpOperator` but it won't be clean unless we pass in another boolean to signal before or after behavior since the static inner classes will make it a bit ugly to use abstract methods. wdyt?

Reply:
I consolidated it, because it seems better. We can revisit if we don't like it.